test/AutoDiff/compiler_crashers_fixed/sr12559-derivative-attr-type-checking-parse-stdlib.swift

```diff
@@ -0,0 +1,30 @@
+// RUN: not %target-swift-frontend-typecheck -parse-stdlib %s
+
+// SR-12559: `@derivative` attribute type-checking crash.
+// This program is not valid, but the compiler should not crash nonetheless.
+// Reproducible only with `-parse-stdlib`.
+
+// The crash occurs because it is not sufficient for attribute type-checking to
+// check for `Differentiable` conformances. We must also checking for invalid
+// associated types:
+//
+//   (lldb) p valueResultConf.dump()
+//   (normal_conformance type=AnyDerivative protocol=Differentiable
+//     (assoc_type req=TangentVector type=<<error type>>))
+
+import _Differentiation
+
+struct AnyDerivative: Differentiable {
+  init<T>(_ base: T) {}
+
+  @derivative(of: init)
+  static func _vjpInit<T: Differentiable>(
+    _ base: T
+  ) -> (value: AnyDerivative, pullback: (AnyDerivative) -> T.TangentVector) {
+    fatalError()
+  }
+
+  typealias TangentVector = AnyDerivative
+}
+
+// Assertion failed: (resultTan && "Original result has no tangent space?"), function getAutoDiffDerivativeFunctionLinearMapType, file /Users/danielzheng/swift-merge/swift/lib/AST/Type.cpp, line 5190.
```

Comment:
An alternative fix could be to change `AnyFunctionType::getAutoDiffDerivativeFunctionLinearMapType` (and similar functions) to return `Optional<AnyFunctionType *>` instead of asserting `false` on failure.

This changes invariants and requires more work though. For diagnostic purposes, we may need sth like `Expected<AnyFunctionType *>` so the failure case can contain info about which parameter or result is non-`Differentiable`.